content/blog/2015-06-17-practical-anomaly-detection.md

```diff
@@ -0,0 +1,130 @@
+---
+title: Practical Anomaly Detection
+created_at: 2015-06-17
+kind: article
+author_name: Brian Brazil
+---
+
+In his *[Open Letter To Monitoring/Metrics/Alerting Companies](http://www.kitchensoap.com/2015/05/01/openlettertomonitoringproducts/)*,
+John Allspaw asserts that attempting "to detect anomalies perfectly, at the right time, is not possible".
+
+I have seen more than one attempt by talented engineers to build a system to
+automatically detect and diagnose problems based on time series data. While it
+is certainly possible to get a demonstration working, the data always turned
+out to be too noisy to make this approach work for anything but the simplest of
+real-world systems.
+
+All hope is not lost though. There are some common anomalies which you can
+detect and deal with. The Prometheus [query
+language](../../../../../docs/querying/basics/) gives you the tools to discover
+these while avoiding false positives.
+
+## Building a query
+
+A common problem is a small number of servers within a service not
+performing as well as the rest, such as by having increased latency.  While
+the issue may not impact your overall SLA, it is worth having a human look when
+they have a chance.
+
+Let us say that we have `instance:latency_seconds:mean5m` representing the
+average query latency for each instance of a service, calculated from a
+[Summary](../../../../../docs/concepts/metric_types/#summary) metric.
+
+A simple way to start would be to look for instances with latency
+a more than two standard deviations above the mean:
+
+```
+  instance:latency_seconds:mean5m
+> on (job) group_left(instance)
```

Comment:
The one visual issue I have with this indentation style is that when using 2-space indents and having a single-character filter operator (`>`), the `on` clause is at the same indentation level as the operands, so visually it can easily be mistaken as part of the operand (RHS). Hmm. 4-space indents would make that clearer:

```
    instance:latency_seconds:mean5m
> on (job) group_left(instance)
    (
        avg by (job)(instance:latency_seconds:mean5m) 
      + on (job)
        2 * stddev by (job)(instance:latency_seconds:mean5m)
    )
```

...or is that too much indentation? :) What do you think?